File: apps/python-src/worker.py
import sys
import os 
from taskRunner import simulateOperation
from tracer import Tracer

class Worker:
    def __init__(self, id):
        self.id = id
        
    def load_data(self):
        simulateOperation(
            minDelay=.175, maxDelay=.225, failure=0.1
        )

    def save_data(self):
        simulateOperation(
            minDelay=.380, maxDelay=.700, failure=.2
        )

    def process_data(self):
        simulateOperation(
            minDelay=1.5, maxDelay=2.5
        )

    def execute_task(self):
        tracer = Tracer()
        span = tracer.create_continuation_span(
            span_name='w2-{}'.format(self.id),
            context_id=self.id
        )
        span.set_tag('job-id', self.id)
        span.log_kv({
            'event': 'debug', 
            'message': 'Start execution for job id: {}'.format(self.id)
        })
        try:
            childSpan = tracer.create_span(
                span_name='load-data',
                parent_span=span
            )
            self.load_data()
            childSpan.finish()
            childSpan = tracer.create_span(
                span_name='process-data',
                parent_span=span
            )
            self.process_data()
            childSpan.finish()
            childSpan = tracer.create_span(
                span_name='save-data',
                parent_span=span
            )
            self.save_data()
            childSpan.finish()
        except Exception as e:
            span.set_tag('error', e)
        finally:
            span.log_kv({
                'event': 'debug', 
                'message': 'End executionfor job id: {}'.format(self.id)
            })
            span.finish()
            tracer.flush_spans()

if __name__ == "__main__":
    Worker(id=sys.argv[1]).execute_task()
    simulateOperation(.75)
    sys.exit(0)
    

File: apps/nodejs-src/worker.js
const taskRunner = require(`./taskRunner.js`)
const tracer = require(`./tracer.js`).Tracer(
    process.env.JAEGER_NODEJS_APP_NAME
)
const tags = require('opentracing').Tags

class Worker {
    constructor(id, parentSpan) {
        this.id = id
        this.parentSpan = parentSpan
    }

    async executeHttpRequest(url) {
        const opName = 'http-request'
        const span = tracer.createSpan(opName, this.parentSpan)
        span.setTag('job-id', this.id)
        span.setTag(tags.HTTP_METHOD, url)
        await taskRunner.simulateOperation(500, 2500, 0.3)    
        span.log({'event': opName});
        span.finish()
    }

    async loadData() {
        const opName = 'load-data'
        const span = tracer.createSpan(opName, this.parentSpan)
        span.setTag('job-id', this.id)
        await taskRunner.simulateOperation(100, 150, 0.1)
        const rows = []
        for (let index = Math.random() * 100; index > 0; index--) {
            rows.push(index)
        }
        span.log({
            'event': opName, 
            'rows-count': rows.length, 
            'message': 'Data loaded from db'
        })
        span.finish()
    }

    async processData() {
        const opName = 'process-data'
        const span = tracer.createSpan(opName, this.parentSpan)
        span.setTag('job-id', this.id)
        await taskRunner.simulateOperation(500, 1500)
        const value = Math.random() * 10
        span.log({
            'event': opName, 
            'result-value': value, 
            'message': 'Data processed'})
        span.finish()
        return value
    }

    async saveData() {
        const opName = 'save-data'
        const span = tracer.createSpan(opName, this.parentSpan)
        span.setTag('job-id', this.id)
        await taskRunner.simulateOperation(200, 250, 0.1)
        span.log({
            'event': opName, 
            'message': 'Result correctly saved on db'});
        span.finish()
    }
}

(async (jobId) => {
    const parentSpan = await tracer.createContinuationSpan(
        `w1-${jobId}`, jobId
    ) 
    parentSpan.setTag('job-id', jobId)
    parentSpan.log({
            'event': 'debug', 
            'job-id': jobId, 
            'message': `Start execution for job id: ${jobId}`
    })
    try {
        const worker = new Worker(jobId, parentSpan)
        await worker.executeHttpRequest('https://request-to-url.com/')
        await worker.loadData()
        await worker.processData()
        await worker.saveData()
    } catch(err) {
        parentSpan.setTag(tags.ERROR, true)
        parentSpan.log({
            'event': 'error', 
            'error.object': err, 
            'message': err.message, 
            'stack': err.stack
        })
    }
    parentSpan.log({
        'event': 'debug', 
        'job-id': jobId, 
        'message': `End execution for job id: ${jobId}`
    })
    parentSpan.finish()
    setTimeout(() => {
        process.exit(0)
    }, 1000)
})(process.argv[2])